### PR TITLE
Add tensor quantization

### PR DESCRIPTION
**Pull request by laggui**
### Checklist

- [x] Confirmed that `run-checks all` script has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

Progress towards #464 

### Changes

Support for static per-tensor quantization.
- Added new `DType::QFloat`
- Added backend trait associative type `QuantizedTensorPrimitive`
- Added `QTensorOps` for `quantize`/`dequantize` ops
- Added `TensorPrimitive` enum for float tensors to include float and qfloat (quantized type)
  - All existing float ops now automatically retrieve the `.tensor()` (dequantized) before calling the backend implementation on the `FloatTensorPrimitive`
- Added `QuantizationStrategy` enum with per-tensor affine and symmetric int8 quantization

*Note: QAT support should be added in a future PR*

### Testing

Added unit tests for affine and symmetric per-tensor quantization

### CODE REVIEW

File: crates/burn-tensor/src/tensor/api/autodiff.rs [hunk 1]
```diff
@@ -53,13 +87,13 @@ impl<B: AutodiffBackend> BasicAutodiffOps<B> for Float {
...
     ) -> <Self::InnerKind as TensorKind<<B as AutodiffBackend>::InnerBackend>>::Primitive<D> {
-        B::inner(tensor)
+        TensorPrimitive::Float(B::inner(tensor.tensor()))
     }
 
     fn from_inner<const D: usize>(
         inner: <Self::InnerKind as TensorKind<<B as AutodiffBackend>::InnerBackend>>::Primitive<D>,
     ) -> <Self as TensorKind<B>>::Primitive<D> {
-        B::from_inner(inner)
+        TensorPrimitive::Float(B::from_inner(inner.tensor()))
```
Comment (nathanielsimard) [inline crates/burn-tensor/src/tensor/api/autodiff.rs:96]: We will need to implement a q_inner and q_from_inner at some point.

File: crates/burn-tensor/src/tensor/api/float.rs [hunk 2]
```diff
@@ -119,23 +137,34 @@ where
     /// }
     /// ```
     pub fn int(self) -> Tensor<B, D, Int> {
-        Tensor::new(B::float_into_int(self.primitive))
+        // NOTE: should `tensor.int()` return the quantized values?
```
Comment (nathanielsimard) [inline crates/burn-tensor/src/tensor/api/float.rs:140]: No since we don't support QInt.

**Review by nathanielsimard — APPROVED** [review]
Some comments, but LGTM great job!

File: crates/burn-tensor/src/tensor/api/float.rs (hunk 2 above)
Comment (laggui) [reply crates/burn-tensor/src/tensor/api/float.rs:140]: Whoops, that comment should have been removed 😄

File: crates/burn-tensor/src/tensor/api/autodiff.rs (hunk 1 above)
Comment (laggui) [reply crates/burn-tensor/src/tensor/api/autodiff.rs:96]: Oh, this should be fairly straightforward. I'll add it.

**Comment by laggui** [comment]
Not sure what's up with the wgpu tests, they seem to be failing intermittently even on main

```
error: test failed, to rerun pass `-p burn-wgpu --lib`
  
  Caused by:
    process didn't exit successfully: `/home/runner/work/burn/burn/target/debug/deps/burn_wgpu-c73d9b81b38abf9c --color=always` (signal: 11, SIGSEGV: invalid memory reference)
```

**Comment by nathanielsimard** [comment]
> Not sure what's up with the wgpu tests, they seem to be failing intermittently even on main
> 
> ```
> error: test failed, to rerun pass `-p burn-wgpu --lib`
>   
>   Caused by:
>     process didn't exit successfully: `/home/runner/work/burn/burn/target/debug/deps/burn_wgpu-c73d9b81b38abf9c --color=always` (signal: 11, SIGSEGV: invalid memory reference)
> ```

Working on it

**Pull request merged**